app/src/main.rs

```diff
@@ -10,6 +10,8 @@ fn main() {
     m.log_violation("unauthorized");
     let violations = m.take_violations();
     for v in violations {
+        //let d1:<U:=
+        //let d=v.as_ref();
         println!("{}", v.as_ref());
```

Comment:
@jdm 
I am really stuck at this error in this line. Could you please help?
_________________________________________________________________
 the type of this value must be known in this context
  --> src/main.rs:15:24
   |
15 |         println!("{}", v.as_ref());
   |                        ^^^^^^^^^^


_________________________________________________________________
I have used traits on properties of Logger and LogEntry to remove the dependancy of base on middle crate. Is my approach right or it needs some improvement?

Reply:
Try this:
```rust
let s: &str = v.as_ref();
println!("{}", s);
```

Reply:
Yes I have tried this, still the error remains same.

Reply:
Oh, the problem is that you're calling create_middleware but not providing any concrete types (like `create_middleware::<Foo, Bar>()`), so it's trying to infer them and getting stuck.

Reply:
Oh yes! That was it. Thank you so much for helping me.